### DISCUSSION TITLE
antd5 中怎样在定义了 `colorPrimary` 之后获取计算出的 调色板?

### DISCUSSION THREAD

**Discussion by xsjcTony** [ant-design/ant-design#47056] [Q&A] [answered]
antd5 中怎样在定义了 `colorPrimary` 之后获取计算出的 调色板? (100, 200, 300 ... 900) 各个调色板的值

同时需要考虑当前是 `defaultAlgorithm` 还是 `darkAlgorithm`

**Comment by Wxh16144** [comment]
没懂，你说的 defaultAlgorithm 和 darkAlgorithm 不就是算法具体实现逻辑么？

**Comment by MadCcc** [comment]
https://ant.design/theme-editor-cn?tab=design
![image](https://github.com/ant-design/ant-design/assets/27722486/e70c7a87-c38c-4201-9571-de0dfddb92be)
不知道你指的是不是这个

**Reply by xsjcTony** [reply]
这些有token可以获取，我想要的是这个 
![image](https://github.com/ant-design/ant-design/assets/45784210/40955347-b794-4afd-9d84-c6ef7db88e8a)
没有办法通过 `useToken()` 直接获取吗， 因为token是和当前算法绑定的所以会方便一点 如果有的话

**Reply by Wxh16144** [reply]
`token['blue-2']` 可以获取到，文档没写而已

**Reply by xsjcTony** [reply]
怪不得文档里没找到🤣，那如果定义了自己的 `colorPrimary`, 就是 `token['primary-2']`了?

**Reply by xsjcTony** [reply]
> 怪不得文档里没找到🤣，那如果定义了自己的 `colorPrimary`, 就是 `token['primary-2']`了?

好像没有 TS没定义

**Reply by Wxh16144** [reply]
有的吧🤔？
<img width="1582" alt="image" src="https://github.com/ant-design/ant-design/assets/32004925/de73e0de-5b04-4807-9b3d-ee54613ab63c">

<img width="434" alt="image" src="https://github.com/ant-design/ant-design/assets/32004925/4b90a645-abbb-4eb0-9548-348d370711c5">

**Reply by Wxh16144** [reply]
>  那如果定义了自己的 colorPrimary, 就是 token['primary-2']了?

不是， Primary 就是 blue  ~（不敢肯定，90%可能是这样 2333~ ~

验证了一下并不是

**Reply by xsjcTony** [reply]
并不是🤣问题就在这了， 有办法获取计算出的 Primary 的调色板吗

**Reply by Wxh16144** [reply, marked as answer]
我明白你的意思了，你拿到上下文的中的 `colorPrimary` 然后用 `@ant-design/colors` 提供的 `generate` 方法再计算一遍。生成一个面板。

**Reply by xsjcTony** [reply]
OK 确认一下， 默认就是 `defaultAlgorithm`, 然后 `theme: 'dark'` 就相当于 `darkAlgorithm` 是等价的对吧?

**Reply by xsjcTony** [reply]
自己弄个hook也不算麻烦, 但是我感觉这个其实可以从 `useToken` 暴露出去 (因为反正是计算过的吧?) 应该会有其他人也会用到

**Reply by Wxh16144** [reply]
不过借你这个讨论，再加上之前讨论的算法是一个 pipeline，可以自定义一个算法，把 colorPrimary 的面板算法追加到 CP 的 `theme.algorithm` 里面

**Reply by Wxh16144** [reply]
> 自己弄个hook也不算麻烦, 但是我感觉这个其实可以从 `useToken` 暴露出去 (因为反正是计算过的吧?) 应该会有其他人也会用到

如你所说自己弄一个 hook 也不麻烦，antd 这边暂且就不会考虑加入了[doge]

**Reply by xsjcTony** [reply]
所以其实 `defaultAlgorithm` 和 `@ant-design/colors` 的 `generate` 的默认算法还不一样? 阿巴阿巴. 我还以为是用同样的算法算出来, 然后挑几个给予意义

**Reply by Wxh16144** [reply]
https://github.com/ant-design/ant-design/blob/7bd08ba/components/theme/themes/default/index.ts 具体看算法部分，我也不是很了解。

**Reply by MadCcc** [reply]
把 `blue` 改了色板就变了，另外推荐用 `token.blue1`，`token['token-1']` 不太符合正常书写，会被废弃

**Reply by xsjcTony** [reply]
我用 colors 包自己算吧 改blue不知道会不会有啥其他问题🤣

**Comment by Wxh16144** [comment]
补充一下：

_useAntdPrimary.ts_

```tsx
import { generate } from '@ant-design/colors';
import { ConfigProvider, theme } from 'antd';
import * as React from 'react';

const toArray = <T>(candidate?: T | T[]): T[] => {
  if (candidate === undefined) return [];
  return Array.isArray(candidate) ? candidate : [candidate];
};

function useAntdPrimary() {
  const ctx = React.useContext(ConfigProvider.ConfigContext);
  const { token } = theme.useToken();

  const antdThemeAlgorithm = React.useMemo(() => {
    const { theme } = ctx;
    return toArray(theme?.algorithm).filter(Boolean);
  }, [ctx.theme]);

  const hasDarkTheme = antdThemeAlgorithm.includes(theme.darkAlgorithm);

  const colorPalette = React.useMemo(() => {
    return generate(token.colorPrimary, {
      theme: hasDarkTheme ? 'dark' : 'default',
    });
  }, [hasDarkTheme, token.colorPrimary]);

  return colorPalette;
}

export default useAntdPrimary;

```

_index.tsx_

```tsx
import { Button, ColorPicker, ConfigProvider, theme } from 'antd';
import { useState } from 'react';
import useAntdPrimary from './useAntdPrimary';

function Colors() {
  const primaryPalette = useAntdPrimary();
  const { token } = theme.useToken();

  function renderBlueColor(_, index) {
    return (
      <li key={index} style={{ backgroundColor: token[`blue-${index}`] }}>
        {token[`blue-${index}`]}
      </li>
    );
  }

  function renderPrimaryColor(color, index) {
    return (
      <li key={index} style={{ backgroundColor: color }}>
        {color}
      </li>
    );
  }

  return (
    <div>
      <ul>{Array.from({ length: 10 }, renderBlueColor)}</ul>
      <ul>{primaryPalette.map(renderPrimaryColor)}</ul>
    </div>
  );
}

function App() {
  const [color, setColor] = useState('#1890ff');

  return (
    <>
      <h1>color:{color}</h1>
      <ConfigProvider
        theme={{
          token: {
            colorPrimary: `${color}`,
          },
        }}
      >
        <Button type="primary">Click</Button>
        <Colors />
      </ConfigProvider>
      <ColorPicker value={color} onChange={(v) => setColor(v.toHexString())} />
    </>
  );
}

export default App;

```
